### PR TITLE
[DNM] Test REST sample

### PR DESCRIPTION
**Pull request by tony-le-24**
This is a test sample that is written by Tony Le for learning purpose only, Please do not merge.

### CODE REVIEW

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/Kconfig [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+config REST_DEVICE_MESSAGE_SEND_LED_NUM
+	int "The LED number to use for indicating outgoing REST request"
+	default 1
+
+config REST_DEVICE_MESSAGE_BUTTON_EVT_NUM
+	int "Button number to use for device interactions"
+	default 1
+
+config REST_DEVICE_MESSAGE_DO_JITP
+	bool "Ask to perform just-in-time provisioning (JITP) via REST on startup"
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_device_message_tol2/Kconfig:26]: Since JITP is not supported in your sample, you should remove all mention of it in this file as well as in others, like README.rst.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+void get_date_time_lte(void)
+{
+	char time_buf[64];
+	int err;
+
+	LOG_INF("Waiting for modem to acquire time from network");
+
+	do {
+		err = nrf_modem_at_cmd(time_buf, sizeof(time_buf), "AT%%CCLK?");
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:87]: I know the original sample does it this way, but there is another way.

Take a look at the date_time library, and the event notifications for things like `DATE_TIME_OBTAINED_NTP` and `DATE_TIME_OBTAINED_MODEM`.

I suggest you try to use that to detect when time is obtained, and synchronize on that instead of the existing busy loop.

**Comment by plskeggs** [comment]
Your commit message should reference the Jira ticket that led to you creating this PR.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c (hunk 2 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:87]: I'll take a look at them and change my code accordingly.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/Kconfig (hunk 1 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/Kconfig:26]: I'll delete all the mentions about JITP. Thank you.

**Comment by tony-le-24** [comment]
I'll edit my commit message as well.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+See  `nRF Cloud Security`_ for more details.
+Generating valid JWTs requires the network carrier to provide date and time to the modem, so the sample must first connect to an LTE carrier and determine the current date and time before REST requests can be sent.
+
+Note also that the `nRF Cloud REST API`_ is stateless.
+This differs from the `nRF Cloud MQTT API`_, which requires you to establish and maintain an `MQTT`_ connection while sending `Device Messages <nRF Cloud Device Messages_>`.
+
+User interface
+**************
+
+Once the device is provisioned and connected, it will periodically send messages to the cloud in 60 seconds.
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst:41]: I would reword this to say "send messages to the cloud once every 60 seconds."

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+Dependencies
+************
+
+This sample uses the following |NCS| libraries:
+
+* :ref:`lib_nrf_cloud`
+* :ref:`lib_nrf_cloud_rest`
+* :ref:`lib_modem_jwt`
+* :ref:`lte_lc_readme`
+* :ref:`dk_buttons_and_leds_readme`
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst:85]: Is the dk_buttons_and_leds library still needed?

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+#include <zephyr/kernel.h>
+#include <modem/nrf_modem_lib.h>
+#include <nrf_modem_at.h>
+#include <modem/modem_info.h>
+#include <zephyr/settings/settings.h>
+#include <net/nrf_cloud.h>
+#include <net/nrf_cloud_rest.h>
+#include <zephyr/sys/reboot.h>
+#include <zephyr/logging/log.h>
+#include <dk_buttons_and_leds.h>
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:16]: Is dk_buttons_and_leds.h still necessary?

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+ */
+
+#include <zephyr/kernel.h>
+#include <modem/nrf_modem_lib.h>
+#include <nrf_modem_at.h>
+#include <modem/modem_info.h>
+#include <zephyr/settings/settings.h>
+#include <net/nrf_cloud.h>
+#include <net/nrf_cloud_rest.h>
+#include <zephyr/sys/reboot.h>
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:14]: Is anything in <zephyr/sys/reboot.h> used?  Maybe you can remove it.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst (hunk 3 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst:41]: Well, I meant to say that the service has a timeout of maximum 60 seconds, you cannot change it to send messages indefinitely. I'll change it to something more reasonable.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst (hunk 4 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst:85]: I don't think so, will remove it.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c (hunk 5 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:16]: I will remove it.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c (hunk 6 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:14]: I will remove that, too.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst (hunk 3 above)
Comment (plskeggs) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst:41]: Hmm.  OK, actually, your main() function has a loop which sends messages until there is an error, once every 3 seconds.

Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst:41]: Yes, the error will happens eventually, since I can set the timeout in rest-context but if it's higher than 60,000ms, it will be set to the value of 60,000 in runtime; also setting it 0 will make the timeout immediately.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c (hunk 6 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:14]: Actually, there are some more unnecessary libraries, I'll remove them as well.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+#define TIME_TO_RETRY 5
+#define MAX_TIME_OUT 60000
+#define REST_RX_BUF_SIZE 2100
+
+char device_id[NRF_CLOUD_CLIENT_ID_MAX_LEN + 1];
+K_SEM_DEFINE(lte_connect, 0, 1);
+K_SEM_DEFINE(modem_time, 0, 1);
+char rx_buf[REST_RX_BUF_SIZE];
+struct nrf_cloud_rest_context rest_context = {
+						   .connect_socket = -1,
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:28]: `.connect_socket` does not look to be aligned with the other elements -- it should be aligned

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst (hunk 3 above)
Comment (plskeggs) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/README.rst:41]: The 60 seconds is used in case the server does not respond to the REST call, but normally it will respond to every call within a few seconds.  I don't feel it is useful to say "in 60 seconds."

The "happy case" is that a message will be sent every 3 seconds forever.

**Review by plskeggs — COMMENTED** [review]
It looks good enough now, since the goal is not to merge this but just learn a bit about NCS, pull requests, and code reviews.

File: samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c (hunk 7 above)
Comment (tony-le-24) [reply samples/nrf9160/nrf_cloud_rest_device_message_tol2/src/main.c:28]: Ohh I forgot to edit the code after using clang-format, it's a weird alignment every time I used.

**Comment by tony-le-24** [comment]
Since all the tests are green (except for the skipped/aborted ones), and the goals are met, I'll close this PR.
